### PR TITLE
SQL compound statement awareness in ScriptUtils

### PR DESCRIPTION
**Pull request by rnorth**
This came out of #570. In brief, for DB init script handling, `ScriptUtils` is not capable of handling scripts that contain compound statements in stored procedures (`BEGIN...END` sections). It will simply split the script on semicolons, mid-procedure, thus breaking the script. 

An example would be:
```
CREATE PROCEDURE count_foo()
    BEGIN
      SELECT *
        FROM bar;   -- <-- ScriptUtils will split here, which is wrong
      SELECT 1
        FROM dual;
    END; -- <-- statement should end here
```

This prevents users from having stored procedures in their init scripts, which feels like an unpleasant limitation.

The Spring developers behind `ScriptUtils` chose not to go down the path of parsing the SQL, and I can see why they would do this. For arbitrary scripts and many SQL dialects, this could be quite open ended.

However, I wonder if we can still improve on this _just enough to be useful_ with a small change, so have created this WIP PR. I'm not intending for this to be the solution to all possible problems, but hope it unlocks us in some ways.

As-is, this: 
* won't split statements in between BEGIN/END tokens
* can deal with nested blocks
* seems to be OK with BEGIN/END appearing in string literals or comments

I don't believe this could create any regressions in its current form, and I believe BEGIN/END support is all we need to improve things greatly.

### CODE REVIEW

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 1]
```diff
@@ -185,6 +192,21 @@ else if (c == ' ' || c == '\n' || c == '\t') {
...
+			return false;
+		}
+
+		for (int i = 0; i < needleChars.length; i++) {
+			if (Character.toLowerCase(content[offset + i]) != Character.toLowerCase(needleChars[i])) {
+				return false;
+			}
+		}
+		return true;
+	}
```
Comment (rnorth) [inline modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: Performance probably isn't optimal; just aiming for clarity right now.
`needle` parameter is the substring we're looking for - I'll rename to be clearer.

Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: Changed

**Comment by jeacott1** [comment]
I don't think this fixes the problem, see my comment here:
https://github.com/testcontainers/testcontainers-java/issues/608#issuecomment-373326735

**Comment by rnorth** [comment]
@bsideup @kiview I'd like to resurrect this PR, as it is a quick win to enable better SQL init scripts for JDBC containers. Issues such as #946 and #570 will continue to catch people out, which is silly when we have a solution that works well enough for typical usage.

Please could you review?

For reference [init_mysql.sql](https://github.com/testcontainers/testcontainers-java/blob/424141b0234fd0d91f4f83cc34c6ed32d1422e18/modules/jdbc-test/src/test/resources/somepath/init_mysql.sql) shows the kinds of syntax that this PR enables, and is included in this PR for testing.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 2]
```diff
@@ -185,6 +194,21 @@ else if (c == ' ' || c == '\n' || c == '\t') {
...
 	}
 
+	private static boolean contentMatches(char[] stringChars, int offset, String substring) {
+		final char[] substringChars = substring.toCharArray();
+		final int end = offset + substringChars.length;
+		if (stringChars.length < end) {
+			return false;
+		}
+
+		for (int i = 0; i < substringChars.length; i++) {
```
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: Done - it actually could be simplified much further!

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 3]
```diff
@@ -137,7 +138,15 @@ public static void splitSqlScript(String resource, String script, String separat
...
 				inLiteral = !inLiteral;
 			}
-			if (!inLiteral) {
+			if (!inLiteral && contentMatches(script, i, "BEGIN")) {
+				compoundStatementDepth++;
+			}
+			if (!inLiteral && contentMatches(script, i, "END")) {
+				compoundStatementDepth--;
+			}
+			final boolean inCompoundStatement = compoundStatementDepth != 0;
```
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: Good point! Done.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 4]
```diff
@@ -189,6 +198,13 @@ else if (c == ' ' || c == '\n' || c == '\t') {
 		}
 	}
 
+	private static boolean contentMatches(String string, int offset, String substring) {
```
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: Yeah, I'm happier with this name so have changed.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 5]
```diff
@@ -137,7 +138,15 @@ public static void splitSqlScript(String resource, String script, String separat
 			if (c == '\'') {
```
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: I've now modified the code to handle string delimiters better, and tested with some fairly nasty permutations of input.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rnorth) [reply modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]: I've reverted back to a local boolean - given the additional changes made, it's become cleaner again, I think.

File: modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java [hunk 6]
```diff
@@ -134,10 +138,39 @@ public static void splitSqlScript(String resource, String script, String separat
...
+				} else if (currentLiteralDelimiter == c) { // find end of string literal
+					currentLiteralDelimiter = null;
+				}
+			}
+			final boolean inLiteral = currentLiteralDelimiter != null;
+
+			if (!inLiteral && containsSubstringAtOffset(script, commentPrefix, i)) {
+				inLineComment = true;
+			}
+			if (inLineComment && c=='\n') {
```
Comment (bsideup) [inline modules/database-commons/src/main/java/org/testcontainers/ext/ScriptUtils.java]:
```suggestion
 			if (inLineComment && c == '\n') {
```

**Comment by rnorth** [comment]
Released in [1.10.4](https://github.com/testcontainers/testcontainers-java/releases/tag/1.10.4)!

**Comment by rnorth** [comment]
Released in [1.10.4](https://github.com/testcontainers/testcontainers-java/releases/tag/1.10.4)!

**Comment by manikmagar** [comment]
Thanks @rnorth, we ran into this issue with postgres a few weeks back.

**Comment by johnhenaot** [comment]
Hi guys, I'm having some issues when using IF statement, it seems that it is assuming the word END in the "END IF" as the END of the BEGIN statement.

`CREATE OR REPLACE PROCEDURE updatealladvertisers()
LANGUAGE plpgsql AS
$$
BEGIN
    IF (SELECT count(1) FROM information_schema.tables WHERE table_schema = 'public' AND table_name='all_advertisers') = 0 THEN
        CREATE TABLE IF NOT EXISTS public.all_advertisers
        (
            partnerid   BIGINT,
            partnername VARCHAR(255),
            partner     VARCHAR(255),
            gsm_id      BIGINT
        );
    END IF;
END $$;`

**Pull request merged**